### DISCUSSION TITLE
RLS is blocking my cron job?

### DISCUSSION THREAD

**Discussion by salo18** [supabase/supabase#23136] [Questions] [answered]
Hi team,
I have a cron job hosted on vercel that hits /api/cron. The cron job works when I hit the endpoint manually and it works when I have RLS disabled. For some reason, I enable RLS and it no longer works.

The database should be updating if certain conditions are met and I created an RLS policy that allows users to update. 

Is there anything about hosting a cron job on vercel that would somehow cause the RLS policy to fail?

Here is my RLS policy

create policy "Enable delete for links based on user_id"
on "public"."links"
for delete using (
  (select auth.uid()) = user_id
);

And here is the api route

`
export const dynamic = 'force-dynamic'
import scrapeOnInsert from "@/app/actions/scrapeOnInsert";

import { NextRequest, NextResponse } from "next/server";
import { createServerSupabaseClient } from "@/libs/supabaseServerClient";

export async function GET(request: NextRequest, response: NextResponse) {
  try {
    // Initialize Supabase client
    const supabase = createServerSupabaseClient();

    // Fetch items from the Supabase table
    const { data } = await supabase
      .from("links")
      .select("*")
      .eq("sent_restock_message", false)
      .order("created_at", { ascending: true });

    for (const link of data) {
      const product: Product = (await scrapeOnInsert(link.url)) as Product;

      if (!product.in_stock) {
        const { error } = await supabase
          .from("links")
          .update({ last_updated: new Date() })
          .eq("link_id", link.link_id);
      } else {
        const { error } = await supabase
          .from("links")
          .update({
            last_updated: new Date(),
            sent_restock_message: true,
            in_stock: true,
          })
          .eq("link_id", link.link_id);
      }
    }
    
    const headers = {
      "Cache-Control": "no-cache, no-store, must-revalidate",
    };

    return new NextResponse(
      JSON.stringify({ message: "Cron job executed successfully" }),
      {
        status: 200,
        headers,
      }
    );

  } catch (error) {
    console.error("Error:", error);
    return NextResponse.json({ error: "error!" }, { status: 500 });
  }
}
`
Any help would be appreciated! I contacted Vercel and they confirmed that the cron job is executing correctly so the issue has to be with Supabase. Since the cron job works when RLS is disabled, I think it has to do with the RLS policy. I just don't understand why RLS would block a process from being done automatically if it works when done manually.

Would it be better to execute this code with pg_cron? https://supabase.com/docs/guides/database/extensions/pg_cron#invoke-supabase-edge-function-every-minute 

Thanks so much for your help

**Comment by GaryAustin1** [comment]
If you are not meeting that RLS then you don't have a user session when you make the call, or your authorization header is not being send with a user jwt.

But you show a delete policy.  None of your calls seem to be delete....  You need an update and select policy met to update.

**Comment by salo18** [comment]
Hi Gary, 

Thanks so much for your help!

I forgot to mention that I do have policies created for select, insert and update as well. 

`create policy "Users can insert."
on public.links for insert
to authenticated                        
with check ( auth.uid() = user_id );   
`

`create policy "Users can read their own data."
on public.links for select
to authenticated                        
with check ( auth.uid() = user_id );   `

I run the cron job on Vercel with RLS off and it works. Then I enable RLS and the cron job runs but the tables are not updated. I tried using this update policy but it was blocking the cron job:
`create policy "Allow users to update their own data" on public.links
for update
using (user_id = auth.uid());`

Then I tried to let anyone make updates and it is still not working:
`CREATE POLICY update_policy ON public.links
FOR UPDATE
TO public
USING (true);`

I am meeting RLS when I hit the endpoint manually but I don't understand why the cron job would cause the RLS conditions to fail. 

How do I ensure the authorization header is sent with the user jwt? I honestly don't understand too much about which settings are needed in the authorization header but this is what I cobbled together:
`  const headers = {
      "Cache-Control": "no-cache, no-store, must-revalidate",
    };

    return new NextResponse(
      JSON.stringify({ message: "Cron job executed successfully" }),
      {
        status: 200,
        headers,
      }
    `

Thanks so much Gary!

**Comment by salo18** [comment, marked as answer]
Hi Gary, I was able to get this to work by using the service role key. I didn't see this anywhere in the documentation but someone else guided me to the solution.

Here is the code in case anyone gets stuck in the future:

import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { cookies } from 'next/headers'
export async function GET(request: NextRequest, response: NextResponse) {
  const cookieStore = cookies()

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.SUPABASE_SERVICE_ROLE_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value
        },
      },
    }
  )

... the rest of your function

I appreciate your help!

**Reply by ebinabo** [reply]
Came here 1 year + later and your answer worked for me as well. Thank you.
